packages/drivers/odsp-driver/src/prefetchLatestSnapshot.ts

```diff
@@ -0,0 +1,207 @@
+/*!
+ * Copyright (c) Microsoft Corporation and contributors. All rights reserved.
+ * Licensed under the MIT License.
+ */
+
+import { v4 as uuid } from "uuid";
+import { assert } from "@fluidframework/common-utils";
+import { ITelemetryBaseLogger } from "@fluidframework/common-definitions";
+import { IResolvedUrl } from "@fluidframework/driver-definitions";
+import {
+    ICacheEntry,
+    IOdspResolvedUrl,
+    IPersistedCache,
+    ISnapshotOptions,
+    OdspResourceTokenFetchOptions,
+    snapshotKey,
+    TokenFetcher,
+} from "@fluidframework/odsp-driver-definitions";
+import { ChildLogger, PerformanceEvent, TelemetryLogger } from "@fluidframework/telemetry-utils";
+import { LocalPersistentCache } from "./odspCache";
+import { ISnapshotCacheValue } from "./odspDocumentStorageManager";
+import {
+    createOdspLogger,
+    fetchAndParseAsJSONHelper,
+    getOdspResolvedUrl,
+    getWithRetryForTokenRefresh,
+    IOdspResponse,
+} from "./odspUtils";
+import { IOdspSnapshot } from "./contracts";
+import { evalBlobsAndTrees, toInstrumentedOdspTokenFetcher } from "./odspUtils2";
+
+/**
+ * Function to prefetch the snapshot and cached it in the persistant cache, so that when the container is loaded
+ * the cached latest snapshot could be used and removes the network call from the critical path.
+ * @param resolvedUrl - Resolved url to fetch the snapshot.
+ * @param getStorageToken - function that can provide the storage token for a given site. This is
+ *  is also referred to as the "VROOM" token in SPO.
+ * @param persistedCache - Cache to store the fetched snapshot.
+ * @param logger - Logger to have telemetry events.
+ * @param hostSnapshotFetchOptions - Options to fetch the snapshot if any. Otherwise default will be used.
+ * @returns - True if the snapshot is cached, false otherwise.
+ */
+export async function prefetchLatestSnapshot(
+    resolvedUrl: IResolvedUrl,
+    getStorageToken: TokenFetcher<OdspResourceTokenFetchOptions>,
+    persistedCache: IPersistedCache = new LocalPersistentCache(),
+    logger: ITelemetryBaseLogger,
+    hostSnapshotFetchOptions: ISnapshotOptions | undefined,
+): Promise<boolean> {
+    const odspLogger = createOdspLogger(ChildLogger.create(logger, "PrefetchSnapshot"));
+    const odspResolvedUrl = getOdspResolvedUrl(resolvedUrl);
+    const snapshotUrl = odspResolvedUrl.endpoints.snapshotStorageUrl;
+    let snapshotOptions: ISnapshotOptions = {
+        deltas: 1,
+        channels: 1,
+        blobs: 2,
+    };
+
+    if (hostSnapshotFetchOptions) {
+        snapshotOptions = {...snapshotOptions, ...hostSnapshotFetchOptions};
```

Comment:
By the way, can the options really be merged like that, or could it sometimes produce a weird state? I.e., if the caller provided options, are you sure you want to merge them with the default ones and that it won't get someone into a weird state?

Reply:
It's fine to merge it that way (we do same in two other places). But I thin Marcelo mentioned some time ago that he was changing defaults and these values are defaults now on server side. @jatgarg, any chance you can run it by Marcelo and see what's current state here?

Reply:
Yes, I will ask.

Reply:
Yes, server have default options when not supplied and as I checked it works and server returns everything.